hotshot-state-prover/src/service.rs

```diff
@@ -423,6 +453,148 @@ pub async fn sync_state<ApiVer: StaticVersionType>(
+    let wallet = EthereumWallet::from(state.config.signer.clone());
+    let provider = ProviderBuilder::new()
+        .wallet(wallet)
+        .on_http(state.config.provider_endpoint.clone());
+
+    tracing::info!(
+        ?light_client_address,
+        "Start syncing light client state for provider: {}",
+        state.config.provider_endpoint,
+    );
+
+    let blocks_per_epoch = state.config.blocks_per_epoch;
+    let epoch_start_block = state.config.epoch_start_block;
+
+    let (contract_state, st_state) = read_contract_state(&provider, light_client_address).await?;
+    tracing::info!(
+        "Current HotShot block height on contract: {}",
+        contract_state.block_height
+    );
+
+    // Update the prover service state if necessary
+    let contract_epoch = epoch_from_block_number(contract_state.block_height, blocks_per_epoch);
+    if contract_epoch != state.epoch {
+        state
+            .sync_with_epoch(contract_epoch)
+            .await
+            .map_err(ProverError::NetworkError)?;
+    }
+
+    let bundle = fetch_latest_state(relay_server_client).await?;
+    tracing::info!("Bundle accumulated weight: {}", bundle.accumulated_weight);
+    tracing::info!("Latest HotShot block height: {}", bundle.state.block_height);
+
+    if contract_state.block_height >= bundle.state.block_height {
+        tracing::info!("No update needed.");
+        return Ok(());
+    }
+    tracing::debug!("Old state: {contract_state:?}");
+    tracing::debug!("New state: {:?}", bundle.state);
+
+    if bundle.state.block_height > epoch_start_block {
+        // We need an epoch catchup to process the current bundle.
+        let bundle_epoch = epoch_from_block_number(bundle.state.block_height, blocks_per_epoch);
+        if bundle_epoch > state.epoch {
+            tracing::info!("Catching up from epoch {contract_epoch} to epoch {bundle_epoch}...");
+            epoch_catchup(
+                state,
+                &provider,
+                light_client_address,
+                st_state,
+                &proving_key,
+                contract_epoch,
+                bundle_epoch,
+            )
+            .await?;
+        }
+
+        // If we reached the epoch root, proceed to the next epoch and ignore the rest bundles from the same epoch
+        if is_ge_epoch_root(bundle.state.block_height as u64, blocks_per_epoch) {
+            tracing::info!("Epoch reaching an end, proceed to the next epoch...");
+            epoch_catchup(
+                state,
+                &provider,
+                light_client_address,
+                st_state,
+                &proving_key,
+                bundle_epoch,
+                bundle_epoch + 1,
+            )
+            .await?;
+            return Ok(());
+        }
+    }
+
+    // Stake table update is already handled in the epoch catchup
+    let (proof, public_input) = generate_proof_helper(
+        state,
+        &bundle.state,
+        &st_state,
+        &st_state,
```

Comment:
is this right? shouldn't the second be `next_st_state`?

Reply:
If it's pre-epoch upgrade, or the block number hasn't reached epoch root. We are not change the next_st_state.